zipkin-common/src/main/scala/com/twitter/zipkin/common/Endpoint.scala

```diff
@@ -20,23 +20,18 @@ import java.net.{InetAddress, InetSocketAddress}
 import java.nio.ByteBuffer
 
 import com.google.common.collect.ComparisonChain
-
-import scala.util.hashing.MurmurHash3
+import com.twitter.zipkin.util.Util._
 
 /**
  * Represents the client or server machine we traced.
  *
  * @param ipv4 ipv4 ip address.
  * @param port note that due to lack of unsigned integers this will wrap.
- * @param _serviceName the service this operation happened on, in lowercase
+ * @param serviceName the service this operation happened on, in lowercase
  */
-// This is not a case-class as we need to enforce serviceName as lowercase
-class Endpoint(val ipv4: Int, val port: Short, _serviceName: String) extends Ordered[Endpoint] {
-
-  /**  the service this operation happened on, in lowercase. */
-  val serviceName: String = _serviceName.toLowerCase
-
-  override lazy val toString = "%s:%d(%s)".format(getHostAddress, port, serviceName)
+case class Endpoint(ipv4: Int, port: Short, serviceName: String) extends Ordered[Endpoint] {
+  checkArgument(serviceName.toLowerCase == serviceName,
```

Comment:
@danchia you mean "require" instead of this? mind giving it a go? Note, we do this a few places, so you could do a reference hunt for Util.checkArgument